### PR TITLE
[BottomNavigation] Add multi-line title support.

### LINKED ISSUE #6955: [BottomNavigation] Support multi-line titles
## Feature Request/Suggestion
Material Guidelines for Bottom Navigation now *discourage* wrapping title text, but no longer forbid it.  We should update Bottom Navigation to support multi-line layout for titles. 

## Acceptance Criteria
* The default behavior is single-line titles with truncation.
* An API exists on `MDCBottomNavigationBar` that allows clients to enable text wrapping.
* If the size of the text is such that the item no longer fits within the safe area of the BottomNavigationBar, the `intrinsicContentSize` and `sizeThatFits:` values reflect the necessary size to fit all of the content.
* If the size of the BottomNavigationBar don't permit all of the content to fit within the bounds of the view, the content is "pinned" to the top of the bar and may extend out of the safe area or beyond the bottom of the screen.
* Snapshot tests covering all edge cases
    * Too large with sufficient bar height
    * Too large with too-short bar height
    

Screen Shot 2019-03-20 at 11 40 58 AM

---
#### Internal data
- Associated internal bug: [b/128997517](http://b/128997517)

### PR DESCRIPTION
**Pull request by romoore**
Adding support for multi-line titles when the icon and text are vertically stacked.

Closes #6955

### CODE REVIEW

File: components/BottomNavigation/src/private/MDCBottomNavigationItemView.h [hunk 1]
```diff
@@ -28,6 +28,13 @@
 @property(nonatomic, copy) NSString *badgeValue;
 @property(nonatomic, copy) NSString *title;
 @property(nonatomic, strong) UIFont *itemTitleFont UI_APPEARANCE_SELECTOR;
+
+/**
+ The number of lines available for rendering the title of this item.
+
+ @note This property is only used when @c titleBelowIcon is @c true.
```
Comment (yarneo) [inline components/BottomNavigation/src/private/MDCBottomNavigationItemView.h:35]: can you clarify that the default is 1 here?

File: components/BottomNavigation/src/private/MDCBottomNavigationItemView.m [hunk 2]
```diff
@@ -577,6 +584,16 @@ - (void)setTitlePositionAdjustment:(UIOffset)titlePositionAdjustment {
   }
 }
 
+- (void)setTitleNumberOfLines:(NSInteger)titleNumberOfLines {
+  _titleNumberOfLines = titleNumberOfLines;
+  self.label.numberOfLines = self.titleBelowIcon ? titleNumberOfLines : 1;
```
Comment (yarneo) [inline components/BottomNavigation/src/private/MDCBottomNavigationItemView.m]: there would be cases the numberOfLines for the label would not be in sync with the titleNumberOfLines iVar. Is this the wanted behavior? If so, can you please clarify this in documentation. Lastly, why use a backing store to begin with and not just use self.label.numberOfLines?

**Review by yarneo — CHANGES_REQUESTED** [review]

File: components/BottomNavigation/src/private/MDCBottomNavigationItemView.h (hunk 1 above)
Comment (romoore) [reply components/BottomNavigation/src/private/MDCBottomNavigationItemView.h:35]: Done.

File: components/BottomNavigation/src/private/MDCBottomNavigationItemView.m (hunk 2 above)
Comment (romoore) [reply components/BottomNavigation/src/private/MDCBottomNavigationItemView.m]: Yes, the expected behavior would be for `titlesNumberOfLines` to apply only when the title and icon are vertically "stacked". That is the only example in the Material Guidelines article showing multi-line titles.  I improved the header docs (it was missing in the public property).  

I felt that the combination of tests captured the behavior, both snapshot and unit tests.  Do you think there is something more that can be done to make the behavior more explicit/obvious?

Comment (romoore) [reply components/BottomNavigation/src/private/MDCBottomNavigationItemView.m]: I forgot to answer the last question.  Because the value of `numberOfLines` on the UILabel is adjusted depending on the layout, a separate variable is required to store the user data.

**Review by codeman7 — APPROVED** [review]

File: components/BottomNavigation/src/private/MDCBottomNavigationItemView.m [hunk 3]
```diff
@@ -577,6 +584,16 @@ - (void)setTitlePositionAdjustment:(UIOffset)titlePositionAdjustment {
...
+}
+
+- (void)setTitleBelowIcon:(BOOL)titleBelowIcon {
+  _titleBelowIcon = titleBelowIcon;
+  self.label.numberOfLines = self.titleBelowIcon ? self.titleNumberOfLines : 1;
```
Comment (jverkoey) [inline components/BottomNavigation/src/private/MDCBottomNavigationItemView.m]: Please extract this repeated line (589 being the other) to a private method.

**Review by jverkoey — APPROVED** [review]

**Review by yarneo — APPROVED** [review]

File: components/BottomNavigation/src/private/MDCBottomNavigationItemView.m (hunk 3 above)
Comment (romoore) [reply components/BottomNavigation/src/private/MDCBottomNavigationItemView.m]: Done
